### ISSUE TITLE
Error in impMean(): Missing data to extrapolate species pairwise distances from closest neighbors

### ISSUE THREAD

**Issue by MaelDore** [damiendevienne/phylter#4]
Hi,

As discussed with Damien, I am posting here an issue I encountered regarding the extrapolation of species pairwise distances from closest neighbors in the absence of available data (i.e., shared gene trees).

While trying to use phylter on my own dataset, I ran into this generic error: “Error in mat[sp1, closest2sp2] : subscript out of bounds”.
It happened that this error is caused during the process of inferring mean pairwise distances for pair of species which do not occur in any shared gene trees (i.e., initial distance matrices).

Let’s say the “rare” species A and B do not share any gene tree. The current strategy, as far as I understood, is to use the mean of the pairwise distances available between the closest neighbors of species A and B, and the closest neighbors of B and A, who share at least one gene tree with the other member of the pair.

However, in my case, I have species with very few occurrences (a dozen of genes) lost in a dataset with species comprising ten of thousands of genes. I happened to run into a particular case where some pairs of rare species A & B do not have neighbors (i.e., the set of species they share a tree with) that actually share a tree with the other member of the problematic pair, causing the error. 
In practice, the while loop in the function ImputeFromClosestNeighbors() never stop because all tries to find a neighbor of species A with an available distance to B fail, leading to an out-of-bonds error when i increase beyond the length of the vector holding the names of all available neighbors of species A (i.e., all species sharing at least one tree with A).

I would like to suggest here some possible fixes to this issue:

1.  Make the error informative
       - Since this error is cause when i reach the full length of the vector of available neighbors, it is possible to detect this specific case and display an informative error inviting the user to remove one (or the two) rare species causing it from its dataset prior to the analyses

2. Offer a systematic solution:
	
    2.1/ Remove sequentially the rare species that cause the error
      - A simple fix to the error that can be implemented systematically would be to remove automatically the rarest species of the pair (i.e., the one having the less available neighbors = sharing gene trees with the fewest species), and run again the pairwise distance inference. A while loop could be implemented to sequentially remove erroneous species (with a warning message), until no pair of species causes this type of error.
	
    2.2/ Infer pairwise distances from closest neighbors of higher degree

      -  A more complex option to save all species in the analyses could be to look for neighbors of higher degree (i.e., neighbors of neighbors of …) with available pairwise distances to infer a distance for the problematic pair of rare species. For instance, if no first degree neighbors of species A could provide an available distance (because none of them share a tree with species B), it could be possible to explore sequentially the neighbors of the closest neighbors of species A (who are not already first degree neighbors of species A) and try to retrieve a distance to B from them. Ultimately, implementing a while loop would allow to look for neighbors of any given degree, to find an available distance to input in the calculation of the missing pairwise distances.

3. Let the user decide
    - Ideally, the choice on the mean to fix this extrapolation issue could be given to the user with an argument in the function to select between the different options :
- 1./ Just show an informative warning inviting to remove the incriminated species from the dataset
- 2.1/ Remove automatically the incriminated species (with a warning) until all pairwise distances can be extrapolated with from 1st degree neighbors for all pair of specie with missing data.
- 2.2/ Automatically look for higher degree neighbors up to N degrees. Or with no limits of neighboring degree to ensure to save all samples in the dataset at any cost.

I hope this feedback will be useful. Thanks for this very nice new phylogenomic tool.

Best,

Maël

**Comment by Menglinhsy** [comment]
Hi, 
I also ran into this generic error: “Error in mat[sp1, closest2sp2]: subscript out of bounds”.

I have 2000 sequences, and it is not easy to check the "rarest species" and delete them, so I prefer the "2.2/ Automatically look for higher degree neighbors up to N degrees. Or with no limits of the neighboring degree to save all samples in the dataset at any cost."

Do you know if this has already been fixed? or is there any way to help to detect the rarest species automatically?

cheers,
Menglin

**Comment by damiendevienne** [comment]
Hi, I need to solve that for sure! 
@Menglinhsy can you send me your input trees so that I can reproduce the bug and correct it? 

damien.de-vienne@univ-lyon1.fr

Thanks a lot
Damien

**Comment by Menglinhsy** [comment]
Hi, Damien,

thank you for the reply.  
here is the input tree.

[tRNA_rRNA_PCG_UCE.txt](https://github.com/damiendevienne/phylter/files/10742559/tRNA_rRNA_PCG_UCE.txt)

cheers,
Menglin

**Comment by damiendevienne** [comment]
Hello Menglin, 
You said you had 2000 sequences, but I only see 4 trees? 
Sorry if I missed something? 
Damien

**Comment by Menglinhsy** [comment]
Hello, Damien,

I am sorry I might not explain the dataset well. I have four trees with around 2000 tips per tree.

cheers,
Menglin

**Comment by damiendevienne** [comment]
Hi, 
This is now solved. In fact I had already implemented this function! 
The error was coming from the fact that I was not treating correctly the case where the tip names were numbers. Tip names were automatically converted to integers when imputing missing values and this was causing a (big) issue.

Now, @Menglinhsy, this does not change the fact that your dataset is not really appropiate for PhylteR: with only 4 trees, it is difficult for PhylteR to find a common signal in the dataset and find outliers that statisticaly fall outside of the distributions... This explains why PhylteR find 0 outliers in your dataset. 

I close this now!
